third_party/amd/lib/TritonAMDGPUToLLVM/MemoryOpToLLVM.cpp

```diff
@@ -44,8 +37,307 @@ class TransLocalLoadOpConversion
+      Type llvmElemTy, ConversionPatternRewriter &rewriter,
+      const ::triton::AMD::TargetInfo &targetInfo) const {
+
+    auto b = TritonLLVMOpBuilder(loc, rewriter);
+    auto *ctx = rewriter.getContext();
+
+    auto S = [ctx](StringRef v) { return StringAttr::get(ctx, v); };
+    auto kReg = S("register");
+    auto kLane = S("lane");
+    auto kWarp = S("warp");
+    auto kBlock = S("block");
+    auto kOffset = S("offset");
+    auto kAddr = S("addr");
+    auto smemPtrTy = ptr_ty(ctx, 3);
+    auto bitWidth = getIntOrFloatOrPtrBitWidth(llvmElemTy);
+
+    if (cvt.hasInDim(kBlock))
+      return failure();
+
+    // Map onto offsets (contiguous part) and addr (non-contiguous part)
+    LinearLayout fullTile;
+    // Contiguous tile
+    LinearLayout tile;
+    // ds_read_tr*_b64 performs a cooperative transposed load across 16
+    // threads. The instruction processes an Nx16 tile (N=4 for 16-bit, N=8 for
+    // 8-bit). The loaded tile is re-packed/transposed where lane i will
+    // receive the i-th column.
+    //
+    // Loaded tile layout (input):     Register layout (output after transpose):
+    //     K0  K1  ... K15               R0  R1  R2  R3
+    // M0[ ............... ]    =>  T0 [ .   .   .   . ]
+    // M1[ ............... ]        T1 [ .   .   .   . ]
+    // M2[ ............... ]        ...
+    // M3[ ............... ]        T15[ .   .   .   . ]
+    //
+    // Each lane loads 64 contiguous bits from LDS. After the transpose,
+    // lane i receives column i from the input (elements strided by 16
+    // the loaded tile).
+    //
+    // For example with N=4 (16-bit):
+    // - Lane 0 receives elements from column 0: originally at [t0,t4,t8,t12]
+    // - Lane 1 receives elements from column 1: originally at [t0,t4,t8,t12]
+    //   These are the second 16 bits loaded by the same lanes before repacking
+    // - Lane 4 receives elements from column 4: originally at [t1,t5,t9,t13]
+    //
+    // Note that there is no restriction on where elements are loaded
+    // from, only that each lane needs to load 64 contiguous bits from shared
+    // memory. We require N number of lanes to be contiguous since they read
+    // consecutive 64 bits loaded from the same lanes.
+    tile = LinearLayout::identity1D(ldsParams.needContigReg, kLane, kOffset);
+
+    const auto isaFamily = targetInfo.getISAFamily();
+    // B8 types on gfx1250 require a different tile with double the contiguity
+    bool doubleB8Contiguity =
+        isaFamily == AMD::ISAFamily::GFX1250 && bitWidth == 8;
+    const unsigned missingLanes =
+        targetInfo.getWarpSize() / tile.getInDimSize(kLane);
+    unsigned otherLanes = 1;
+    if (isaFamily == AMD::ISAFamily::CDNA4) {
+      otherLanes = (bitWidth == 8) ? 2 : 4;
+    } else if (doubleB8Contiguity) {
+      otherLanes = 2;
+    }
+
+    if (doubleB8Contiguity) {
+      fullTile =
+          tile *
+          LinearLayout::identity1D(ldsParams.needContigReg / 2, kReg, kAddr) *
```

Comment:
why is this called "needContigReg"? These registers do not need to be "contiguous", as in "representing a contiguous part of the tensor". We just need the layout to have at least these many registers, which we check in
```
    if (cvt.getInDimSize(kReg) < fullTile.getInDimSize(kReg)) {
      return failure();
    }
```
right?

I guess a better name for this would be `tileSize`?